### PR TITLE
litep2p/discovery: Fix memory leak in `litep2p.public_addresses()`

### PR DESCRIPTION
**Pull request by lexnv**
This PR ensures that the `litep2p.public_addresses()`  never grows indefinitely.
- effectively fixes subtle memory leaks
- fixes authority DHT records being dropped due to size limits being exceeded 
- provides a healthier subset of public addresses to the `/identify` protocol

This PR adds a new `ExternalAddressExpired` event to the litep2p/discovery process.

Substrate uses an LRU `address_confirmations` bounded to 32 address entries. 
The oldest entry is propagated via the `ExternalAddressExpired` event when a new address is added to the list (if capacity is exceeded).

The expired address is then removed from the `litep2p.public_addresses()`, effectively limiting its size to 32 entries (the size of `address_confirmations` LRU).

cc @paritytech/networking @alexggh

### CODE REVIEW

**Review by bkchr — APPROVED** [review]

**Comment by dmitry-markin** [comment]
Why do we have so many (>32) external addresses confirmed? Is there a linked issue in polkadot-sdk?

**Review by skunert — APPROVED** [review]

File: substrate/client/network/src/litep2p/mod.rs [hunk 1]
```diff
@@ -935,6 +935,29 @@ impl<B: BlockT + 'static, H: ExHashT> NetworkBackend<B, H> for Litep2pNetworkBac
...
+						// Litep2p requires the peer ID to be present in the address.
+						let Some(multihash) = litep2p::types::multihash::Multihash::from_bytes(&local_peer_id.to_bytes()).ok() else {
+							continue
+						};
+
+						let address = if !std::matches!(address.iter().last(), Some(Protocol::P2p(_))) {
+							address.with(Protocol::P2p(multihash))
+						} else {
+							address
+						};
```
Comment (dmitry-markin) [inline substrate/client/network/src/litep2p/mod.rs:950]:
```suggestion
						let address = if !std::matches!(address.iter().last(), Some(Protocol::P2p(_))) {
							address.with(Protocol::P2p(local_peer_id.as_ref().clone()))
						} else {
							address
						};
```

**Review by dmitry-markin — APPROVED** [review]

**Comment by lexnv** [comment]
> Why do we have so many (>32) external addresses confirmed? Is there a linked issue in polkadot-sdk?

This issue was reproduced in Rococo on litep2p validators, don't believe we have an gitissue for it. It looks like over time peers report via identify protocol connections on different ports. 
Libp2p has a few events on which we clean up the addresses (AddressExpired / ExpiredListenAddr / ListenerClosed / port-reuse logic).

**Comment by alexggh** [comment]
> Why do we have so many (>32) external addresses confirmed? Is there a linked issue in polkadot-sdk?

Actua

**Comment by dmitry-markin** [comment]
> > Why do we have so many (>32) external addresses confirmed? Is there a linked issue in polkadot-sdk?
> 
> This is the long standing issue for this happening on rococo: [#3519 (comment)](https://github.com/paritytech/polkadot-sdk/issues/3519#issuecomment-1994760856), we never uncovered why so many external addresses accumulate on rococo.

In litep2p backend we have the requirement that at least 2 different peers should report the same external address for it to become confirmed. So, in theory, it should be better with litep2p backend.

**Pull request merged**